airbyte-integrations/connectors/destination-keen/src/main/java/io/airbyte/integrations/destination/keen/KeenRecordsConsumer.java

```diff
@@ -0,0 +1,134 @@
+  }
+
+  @Override
+  protected void startTracked() throws IOException, InterruptedException {
+    projectId = config.get(CONFIG_PROJECT_ID).textValue();
+    apiKey = config.get(CONFIG_API_KEY).textValue();
+    timestampInferenceEnabled = Optional.ofNullable(config.get(INFER_TIMESTAMP))
+        .map(JsonNode::booleanValue)
+        .orElse(true);
+    this.kafkaProducer = KeenDestination.KafkaProducerFactory.create(projectId, apiKey);
+    this.streamNames = getStrippedStreamNames();
+    this.timestampService = new KeenTimestampService(this.catalog, timestampInferenceEnabled);
+    eraseStreams();
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage msg) {
+    if (msg.getType() == Type.STATE) {
+      lastStateMessage = msg;
+      outputRecordCollector.accept(lastStateMessage);
+      return;
+    } else if (msg.getType() != Type.RECORD) {
+      return;
+    }
+
+    final String streamName = getStreamName(msg.getRecord());
+    JsonNode data = this.timestampService.injectTimestamp(msg);
+
+    kafkaProducer.send(new ProducerRecord<>(streamName, data.toString()));
+  }
+
+  private Set<String> getStrippedStreamNames() {
+    return catalog.getStreams()
+        .stream()
+        .map(ConfiguredAirbyteStream::getStream)
+        .map(AirbyteStream::getName)
+        .map(KeenCharactersStripper::stripSpecialCharactersFromStreamName)
+        .collect(Collectors.toSet());
+  }
+
+  private void eraseStreams() throws IOException, InterruptedException {
+    KeenHttpClient keenHttpClient = new KeenHttpClient();
+    LOGGER.info("erasing streams with override options selected.");
+
+    List<String> streamsToDelete = this.catalog.getStreams().stream()
+        .filter(stream -> stream.getDestinationSyncMode() == DestinationSyncMode.OVERWRITE)
+        .map(stream -> KeenCharactersStripper.stripSpecialCharactersFromStreamName(stream.getStream().getName()))
+        .collect(Collectors.toList());
+
+    for (String streamToDelete : streamsToDelete) {
+      LOGGER.info("erasing stream " + streamToDelete);
+      keenHttpClient.eraseStream(streamToDelete, projectId, apiKey);
+    }
+  }
+
+  private String getStreamName(AirbyteRecordMessage recordMessage) {
+    String streamName = recordMessage.getStream();
+    if (streamNames.contains(streamName)) {
+      return streamName;
+    }
+    streamName = KeenCharactersStripper.stripSpecialCharactersFromStreamName(streamName);
+    if (!streamNames.contains(streamName)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
+              Jsons.serialize(catalog), Jsons.serialize(recordMessage)));
+    }
+    return streamName;
+  }
+
+
+  @Override
+  protected void close(boolean hasFailed) {
+    if (!hasFailed) {
+      LOGGER.info("shutting down consumer.");
+      outputRecordCollector.accept(lastStateMessage);
+    }
+    kafkaProducer.close();
```

Comment:
should we close the producer before the last state message was output in case there are records to be flushed and `close()` fails?